### ISSUE TITLE
file watcher unnecessarily triggers compilation when editing (but not saving) Emacs buffers on MS-Windows

### ISSUE THREAD

**Issue by ikappaki** [thheller/shadow-cljs#972]
Hi,

shadow watcher triggers build compilation while editing files in Emacs on MS-Windows. The compilation should be triggered after the file is saved and not while editing.

To reproduce on MS-Windows with latest Emacs (i.e. 27.2)
1. `npx create-cljs-project issue-watch-emacs-ms-win`
2. `cd issue-watch-emacs-ms-win`
3. Add a basic build in `shadow-cljs.edn`
```clojure
;; shadow-cljs configuration
{:source-paths
 ["src/dev"
  "src/main"
  "src/test"]

 :dependencies
 []

 :builds
 {:test
  {:target :browser-test
   :test-dir  "out/test"}}}
```
4. start watching the build with `shadow-cljs watch :test`
```
shadow-cljs - config: d:\clj\issues\issue-watch-emacs-ms-win\shadow-cljs.edn
shadow-cljs - server version: 2.16.10 running at http://localhost:9630
shadow-cljs - nREPL server started on port 53596
shadow-cljs - watching build :test
[:test] Configuring build.
[:test] Compiling ...
[:test] Build completed. (147 files, 1 compiled, 0 warnings, 2.41s)
```
5. In Emacs create an empty `src/main/test.cljs` file and save it.
```
[:test] Compiling ...
[:test] Build completed. (147 files, 1 compiled, 0 warnings, 0.08s)
[:test] Compiling ...
[:test] Build completed. (147 files, 1 compiled, 0 warnings, 0.07s)
```
6. Just type something in the `test.cljs` buffer without saving, a build is unnecessarily triggered
```
[:test] Compiling ...
[:test] Build completed. (147 files, 1 compiled, 0 warnings, 0.07s)
```
This occurs every time you change the buffer after a save.

The reason is that each time you modify a buffer after a save, Emacs on Windows creates a temporary file under `.#<filename>` [to protect against simultaneous editing
](https://www.gnu.org/software/emacs/manual/html_node/emacs/Interlocking.html). In our example Emacs will create a `.#test.cljs` lock file while diting `test.cljs`. Since the lock file ends in .clj[s], is picked up by the watcher and triggers a build. (The same does not occur on Unix because the lock file is a symbolic link rather than a real file.)

Thus a solution could be to exclude certain file name patterns from the watch list (those starting with `.#` in our example). Looking at the code, it appears that  there is only the option to [pass a list of extensions to include](https://github.com/thheller/shadow-cljs/blob/5654aaac86168510837e9b1100f5f25887a0dab6/src/main/shadow/cljs/devtools/server.clj#L403), but there does not appear an interface to exclude some. thus any such endeavour would likely require surgical changes all the way down to the `FileWatcher` facility to support an exclusion pattern? 

Btw, I was expecting a simple `lein install` would create and localy install a shadow-cljs-2.16.10.jar with all the dependencies included, though is missing the `shadow/cljs/dist/babel-worker.js` file. How do you bundle it in the uberjar? I had to copy it over from the official .jar file to make the `shadow-cljs watch` command work while investigating this issue.

Thanks!

**Comment by thheller** [comment]
Actually from what I gather the issue is that it actually creates files and the `:browser-test` is actively looking for files matching the `test$` regexp. So this is in issue in the `:browser-test` handling of new file discovery rather than the FileWatcher itself. That one also properly ignores hidden files which I assume these files are?

**Comment by ikappaki** [comment]
> Actually from what I gather the issue is that it actually creates files and the `:browser-test` is actively looking for files matching the `test$` regexp. So this is in issue in the `:browser-test` handling of new file discovery rather than the FileWatcher itself. That one also properly ignores hidden files which I assume these files are?
> 
> I'm guessing there are no hidden files on windows maybe? I don't have emacs installed so can't promise I get to testing this anytime soon.

Indeed, I've compared a `:browser-test` with a `:browser` :target and the the recompile is triggered only on the first but not the latter. Though, the file event is captured and published on the system bus in both `:target` types, i.e. there is no suppression in the `:browser` case (and further more, I couldn't see anything in the FileWatcher that checks for hidden files, not even sure that even if it did, it would have categorised `.#...` as hidden on MS-Windows).

What is happening and we get a recompilation in the `:browser-test` case is that this build target sets the `:build-options :greedy` flag [and thus falls under the `:else`  condition](https://github.com/thheller/shadow-cljs/blob/5654aaac86168510837e9b1100f5f25887a0dab6/src/main/shadow/cljs/devtools/server/worker/impl.clj#L495), while the `:browser` target does not:

```clojure
(defn do-resource-update
  [{:keys [autobuild last-build-macros last-build-provides build-state] :as worker-state}
   {:keys [namespaces added macros] :as msg}]
 ;; ...
      (cond
        ;; always recompile if the first compile attempt failed
        ;; since we don't know which files it wanted to use
        ;; after that only recompile when files in use by the build changed
        ;; or when new files were added and the build is greedy
        ;; (eg. browser-test since it dynamically adds file to the build)
        (and (pos? (::compile-attempt build-state))
             (not (seq namespaces-used-by-build))
             (not (seq macros-used-by-build))
             (if-not (get-in build-state [:build-options :greedy])
               ;; build is not greedy, not interested in new files
               true
               ;; build is greedy, must recompile if new files were added
               (not (seq added))))
        worker-state

        :else
        (-> worker-state
            ;; only record which namespaces where added/changed
            ;; delay doing actual work until next compilation
            ;; which may not happen immediately if autobuild is off
            ;; the REPL may still trigger compilations independently
            ;; which break if the state is already half cleaned
            (update :namespaces-added set/union added)
            (update :namespaces-modified set/union added namespaces)
            (update :macros-modified set/union macros-used-by-build)
            (cond->
              autobuild
              (build-compile)))))
 ;; ...
)
```

The `npm run build` in `packages/babel-worker` did the trick btw to include the babel-worker.js in the uberjar (though I had to massage the command a little to make it work on windows).

**Comment by ikappaki** [comment]
Hi @thheller,

how do you suggest we take this forward? Just to summarise, greedy builds (such as `:browser-test`) will trigger undesired compilations when Emacs modifies clojure(script) files in memory. This is because Emacs will create an auxiliary `.#<modified filename>` lock file alongside the in memory modified file. The lock file will be picked up by `:greedy` builds which will trigger an unexpected compilation as per the code mentioned in the previous post.

A potential solution is to ignore files starting with `.#` from triggering a new file event, e.g. perhaps in `FileWatcher` or ignoring such an event triggering a compilation, e.g. in the `do-resource-update` fn. I can submit a PR, though what do you think the right solution is?

Thanks

**Comment by thheller** [comment]
I went with just ignoring it directly in the `FileWatcher`. I don't know why emacs doesn't set the hidden flag for those files. I can't think of a reason why any file shadow-cljs should be interested in would start with `.#` so they are now just ignored.

Should be fixed as of `2.16.11`.

**Comment by ikappaki** [comment]
> I went with just ignoring it directly in the `FileWatcher`. I don't know why emacs doesn't set the hidden flag for those files. I can't think of a reason why any file shadow-cljs should be interested in would start with `.#` so they are now just ignored.
> 
> Should be fixed as of `2.16.11`.

Hi @thheller, thanks a lot for the patch! I've noticed that in the case that the file is deeper than the top of the source tree, the event still triggers. I've submitted #975 to cover all cases, feel free to scrutinize!

**Comment by thheller** [comment]
Ah, yeah of course. Fixed in `2.16.12`.

**Comment by ikappaki** [comment]
> Ah, yeah of course. Fixed in `2.16.12`.

Thanks! tested to ignore lock files (at any level) in the `:browser-test` build type!
